### PR TITLE
Prune Node Integration Tests

### PR DESCRIPTION
**Pull request by facs95**
Closes: #XXX

## Description

Integration tests for pruned nodes based on @yihuang tests on cronos."

______

For contributor use:

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: tests/integration_tests/test_pruned_node.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+    )
+    assert tx1 == tx2
+    for name in exp_tx.keys():
+        assert tx1[name] == tx2[name] == exp_tx[name]
+
+    print(
+        w3.eth.get_logs(
+            {"fromBlock": txreceipt.blockNumber, "toBlock": txreceipt.blockNumber}
+        )
+    )
```
Comment (yihuang) [inline tests/integration_tests/test_pruned_node.py:120]: It should assert the logs contains the expect log, I guess I forgot to complete this.

File: tests/integration_tests/configs/pruned_node.jsonnet [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+      pruning: 'everything',
+      'state-sync'+: {
+        'snapshot-interval': 0,
+      },
+    },
+    genesis+: {
+      app_state+: {
+        feemarket+: {
+          params+: {
+            min_gas_multiplier: '0',
```
Comment (yihuang) [inline tests/integration_tests/configs/pruned_node.jsonnet:15]: This parameter should be irrlevant for this test.

File: tests/integration_tests/test_pruned_node.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        "data": "0x000000000000000000000000000000000000000000000000000000000000000a",
+        "transactionIndex": 0,
+        "logIndex": 0,
+        "removed": False,
+    }
+    assert expect_log.items() <= txreceipt.logs[0].items()
+
+    # check get_balance and eth_call don't work on pruned state
+    # we need to check error message here. 
+    # `get_balance` returns unmarshallJson and thats not what it should
```
Comment (yihuang) [inline tests/integration_tests/test_pruned_node.py:72]: The json-rpc error message is not friendly in this case, maybe sth to improve later.

Comment (yihuang) [inline tests/integration_tests/test_pruned_node.py:72]: The json-rpc error message is not friendly in this case, maybe sth to improve later.

File: rpc/backend/evm_backend.go [hunk 4]
```diff
@@ -871,7 +871,7 @@ func (b *Backend) BaseFee(blockRes *tmrpctypes.ResultBlockResults) (*big.Int, er
 	}
 
 	if res.BaseFee == nil {
```
Comment (facs95) [inline rpc/backend/evm_backend.go:873]: I dont think this assumption is accurate. There could be other cases where the `BaseFee` is nill

Comment (facs95) [reply rpc/backend/evm_backend.go:873]: Also, might be good to improve the error message

File: tests/integration_tests/configs/pruned_node.jsonnet (hunk 2 above)
Comment (facs95) [inline tests/integration_tests/configs/pruned_node.jsonnet:15]: We should remove this parameter

**Review by facs95 — COMMENTED** [review]
Thanks @ramacarlucho for helping with the PR 🙏 !  Left a couple of comments for things that I believe should be changed before we merge this

**Comment by yihuang** [comment]
https://github.com/evmos/ethermint/runs/7587033616?check_suite_focus=true

The integration test is not running, I had a fix in another PR: https://github.com/evmos/ethermint/pull/1214

File: rpc/backend/evm_backend.go (hunk 4 above)
Comment (yihuang) [reply rpc/backend/evm_backend.go:873]: Yeah, it could be london hardfork not enabled.

File: rpc/backend/evm_backend.go [hunk 5]
```diff
@@ -879,10 +879,6 @@ func (b *Backend) BaseFee(blockRes *tmrpctypes.ResultBlockResults) (*big.Int, er
 		return nil, err
 	}
 
-	if res.BaseFee == nil {
-		return nil, nil
-	}
```
Comment (yihuang) [inline rpc/backend/evm_backend.go:884]: Why remove this? when london hardfork not enabled, it's `nil`.

File: tests/integration_tests/conftest.py [hunk 6]
```diff
@@ -25,6 +35,8 @@ def cluster(request, ethermint, geth):
         yield ethermint
     elif provider == "geth":
         yield geth
+    elif provider == "pruned":
+        yield pruned
```
Comment (yihuang) [inline tests/integration_tests/conftest.py:39]: For pruned node, the state is not pruned immediately, when the state is not pruned, it should work as a normal node, we only need to test the apis when the state is pruned, so maybe no need to run all the tests on it by default.

File: tests/integration_tests/test_pruned_node.py [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+            HexBytes(b"\x00" * 12 + HexBytes(ADDRS["community"])),
+        ],
+        "data": "0x000000000000000000000000000000000000000000000000000000000000000a",
+        "transactionIndex": 0,
+        "logIndex": 0,
+        "removed": False,
+    }
+    assert expect_log.items() <= tx_receipt.logs[0].items()
+
+    # check get_balance and eth_call don't work on pruned state
```
Comment (yihuang) [inline tests/integration_tests/test_pruned_node.py:57]: Do you removed the negative case for `get_balance`?

File: rpc/namespaces/ethereum/eth/api.go [hunk 8]
```diff
@@ -294,6 +294,11 @@ func (e *PublicAPI) GetBalance(address common.Address, blockNrOrHash rpctypes.Bl
 		return nil, errors.New("invalid balance")
 	}
 
+	// balance can only be negative in case of pruned node
+	if !val.GTE(sdk.NewIntFromUint64(0)) {
```
Comment (yihuang) [inline rpc/namespaces/ethereum/eth/api.go:298]:
```suggestion
	if val.Sign() >= 0 {
```

**Review by yihuang — APPROVED** [review]
LGTM

File: x/evm/keeper/keeper.go [hunk 9]
```diff
@@ -284,6 +284,10 @@ func (k *Keeper) GetNonce(ctx sdk.Context, addr common.Address) uint64 {
 func (k *Keeper) GetBalance(ctx sdk.Context, addr common.Address) *big.Int {
 	cosmosAddr := sdk.AccAddress(addr.Bytes())
 	params := k.GetParams(ctx)
+	// if node is pruned, params is empty. Return invalid value
+	if params.EvmDenom == "" {
```
Comment (fedekunze) [inline x/evm/keeper/keeper.go:288]: We should get the EVM denom directly then using the subspace

**Review by fedekunze — COMMENTED** [review]
See comment. We should only get the EVM denom from the params

File: rpc/namespaces/ethereum/eth/api.go [hunk 10]
```diff
@@ -294,6 +294,11 @@ func (e *PublicAPI) GetBalance(address common.Address, blockNrOrHash rpctypes.Bl
 		return nil, errors.New("invalid balance")
 	}
 
+	// balance can only be negative in case of pruned node
+	if val.IsNegative() {
+		return nil, errors.New("pruned node, cant get balance of pruned states")
```
Comment (fedekunze) [inline rpc/namespaces/ethereum/eth/api.go:299]:
```suggestion
		return nil, errors.New("couldn't fetch balance. Node state is pruned")
```

**Review by fedekunze — APPROVED** [review]

**Review by fedekunze — APPROVED** [review]

**Comment by fedekunze** [comment]
Gosec is prob failing because you're using a fork (?). I'll merge the PR

**Pull request merged**